### PR TITLE
[libunwind] Add GCS support for AArch64

### PR DESCRIPTION
**Pull request by john-brawn-arm**
AArch64 GCS (Guarded Control Stack) is similar enough to CET that we can re-use the existing code that is guarded by _LIBUNWIND_USE_CET, so long as we also add defines to locate the GCS stack and pop the entries from it. We also need the jumpto function to exit using br instead of ret, to prevent it from popping the GCS stack.

GCS support is enabled using the LIBUNWIND_ENABLE_GCS cmake option. This enables -mbranch-protection=standard, which enables GCS. For the places we need to use GCS instructions we use the target attribute, as there's not a command-line option to enable a specific architecture extension.

### CODE REVIEW

File: libunwind/src/cet_unwind.h [hunk 1]
```diff
@@ -35,6 +35,25 @@
...
+// need to guard any use of GCS instructions with __chkfeat though, as GCS may
+// not be enabled.
+#if defined(_LIBUNWIND_TARGET_AARCH64) && defined(__ARM_FEATURE_GCS_DEFAULT)
+#define _LIBUNWIND_USE_CET 1
+#include <arm_acle.h>
+
+#define _LIBUNWIND_POP_CET_SSP(x)                                              \
+  do {                                                                         \
+    if (__chkfeat(_CHKFEAT_GCS)) {                                             \
+      unsigned int tmp = (x);                                                  \
```
Comment (MaskRay) [inline libunwind/src/cet_unwind.h:48]: remove `int`

File: libunwind/src/cet_unwind.h [hunk 2]
```diff
@@ -35,6 +35,25 @@
...
+      while (tmp--) {                                                          \
```
Comment (MaskRay) [inline libunwind/src/cet_unwind.h:49]: drop braces for the single-line simple body

File: libunwind/src/UnwindLevel1.c [hunk 3]
```diff
@@ -170,6 +183,10 @@ unwind_phase1(unw_context_t *uc, unw_cursor_t *cursor, _Unwind_Exception *except
 }
 extern int __unw_step_stage2(unw_cursor_t *);
 
+#if defined(_LIBUNWIND_USE_CET) && defined(_LIBUNWIND_TARGET_AARCH64)
```
Comment (MaskRay) [inline libunwind/src/UnwindLevel1.c:186]: Mixing CET and AArch64 could cause confusion. Could AArch64 use a different variable?

**Comment by john-brawn-arm** [comment]
Ping

**Comment by john-brawn-arm** [comment]
Sorry to pester you about this, but would it be possible to get this approved in time for LLVM19 RC2 on Monday? This is the last piece of GCS work and it would be nice to have it all in a single LLVM release.

**Review by MaskRay — APPROVED** [review]
LGTM. Sorry for the delay.

Q: Is there a way to play with the GCS feature in a VM? Do you have setup instructions?

File: libunwind/src/UnwindLevel1.c [hunk 4]
```diff
@@ -180,8 +197,13 @@ unwind_phase2(unw_context_t *uc, unw_cursor_t *cursor, _Unwind_Exception *except
   // uc is initialized by __unw_getcontext in the parent frame. The first stack
   // frame walked is unwind_phase2.
   unsigned framesWalked = 1;
-#ifdef _LIBUNWIND_USE_CET
+#if defined(_LIBUNWIND_USE_CET)
   unsigned long shadowStackTop = _get_ssp();
+#elif defined(_LIBUNWIND_USE_GCS)
+  unsigned long shadowStackTop = 0;
+  if (__chkfeat(_CHKFEAT_GCS)) {
+    shadowStackTop = (unsigned long)__gcspr();
```
Comment (MaskRay) [inline libunwind/src/UnwindLevel1.c:205]: the brace can be dropped

File: libunwind/src/UnwindLevel1.c [hunk 5]
```diff
@@ -170,6 +183,10 @@ unwind_phase1(unw_context_t *uc, unw_cursor_t *cursor, _Unwind_Exception *except
 }
 extern int __unw_step_stage2(unw_cursor_t *);
 
+#if defined(_LIBUNWIND_USE_GCS)
+// Enable the GCS target feature to permit GCS instructions to be used.
```
Comment (MaskRay) [inline libunwind/src/UnwindLevel1.c:187]: Might be useful to elaborate what instructions need to use the GCS version.

**Review by jinge90 — APPROVED** [review]

**Comment by tru** [comment]
This seems to add a new feature, is it really relevant for a backport?

**Comment by john-brawn-arm** [comment]
> LGTM. Sorry for the delay.
> 
> Q: Is there a way to play with the GCS feature in a VM? Do you have setup instructions?

https://git.yoctoproject.org/meta-arm/log/?h=testing/gcs is a demonstration embedded linux that support GCS, including an emulator capable of running it. https://git.yoctoproject.org/meta-arm/tree/meta-arm-gcs/README?h=testing/gcs is the instructions for building and running it.

**Comment by john-brawn-arm** [comment]
> This seems to add a new feature, is it really relevant for a backport?

This is the last piece of work to enable GCS in the LLVM toolchain (clang and lld support went in a while ago), and I was hoping to get everything in to LLVM 19. It would be rather annoying to have this last part slip to LLVM 20.

The way GCS works is that the `-mbranch-protection=standard` command-line option marks generated objects with the GCS property, and lld then sets the GCS property on the output if all inputs have it, which then indicates to the operating system to enable GCS on that process. Without GCS support in libunwind that means it won't have the GCS property so anything that links with libunwind will have GCS disabled.

Though having said all this, I marked this for LLVM 19 as my reading of https://discourse.llvm.org/t/update-on-llvm-19-x-releases/80511 was that new features are still ok before RC2, but re-reading it it's a bit ambiguous: it says "New features will have to wait until LLVM 20 at this point" and my reading of that was that "this point" means RC2, so new features before RC2 is fine, but if "this point" meant "right now" that means new features aren't OK. If that's the case the it's reasonable for this to not go in RC2.

**Comment by nico** [comment]
Looks like this breaks building on Android: https://ci.chromium.org/ui/p/chromium/builders/try/android-arm64-rel/680348/overview

**Comment by tru** [comment]
> Though having said all this, I marked this for LLVM 19 as my reading of https://discourse.llvm.org/t/update-on-llvm-19-x-releases/80511 was that new features are still ok before RC2, but re-reading it it's a bit ambiguous: it says "New features will have to wait until LLVM 20 at this point" and my reading of that was that "this point" means RC2, so new features before RC2 is fine, but if "this point" meant "right now" that means new features aren't OK. If that's the case the it's reasonable for this to not go in RC2.

I am fine with it going into RC2 if it's part of a on-going work as you explained and that it doesn't create any big issues. Unfortunately with nico's latest comment it seems like this is breaking some build configurations and I am cutting RC2 now, I don't feel comfortable to merge it before that has been investigated. I can probably be convinced to take it in RC3 if it's really helpful to do so and it have been working without any problems.

**Comment by john-brawn-arm** [comment]
> Looks like this breaks building on Android: https://ci.chromium.org/ui/p/chromium/builders/try/android-arm64-rel/680348/overview

Looks like Android here is being built with clang-llvmorg-19-init-14561-gecea8371-3004.tar.xz which looks like it's built from commit https://github.com/llvm/llvm-project/commit/ecea8371 which is after https://github.com/llvm/llvm-project/commit/40d5c2b (which made `-mbranch-protection=standard` enable GCS) but before 3a14ffbae (which added GCS intrinsics to arm_acle.h).

**Comment by nico** [comment]
That sounds believable.

We did update our compiler just an hour or so ago, so it's possible it will work for us now.

However, it suggests that the assumption `defined(__ARM_FEATURE_GCS_DEFAULT)` => arm_acle.h has `__chkfeat()` isn't valid in this snippet in libunwind/src/cet_unwind.h (https://github.com/llvm/llvm-project/pull/99335/files#diff-7c503795ceadc957122b87cbf0dff76624c0576c291268226952dd822c75ea6eR41)

```
#if defined(_LIBUNWIND_TARGET_AARCH64) && defined(__ARM_FEATURE_GCS_DEFAULT)
#define _LIBUNWIND_USE_GCS 1
#include <arm_acle.h>
```

**Comment by john-brawn-arm** [comment]
https://github.com/llvm/llvm-project/pull/101973 for being more careful about enabling GCS.

File: libunwind/src/UnwindRegistersRestore.S [hunk 6]
```diff
@@ -680,7 +680,7 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
   ldr    x16,     [x0, #0x0F8]
   ldp    x0, x1,  [x0, #0x000]  // restore x0,x1
   mov    sp,x16                 // restore sp
-  ret    x30                    // jump to pc
```
Comment (DanielKristofKiss) [inline libunwind/src/UnwindRegistersRestore.S:683]: If the application is compiled with BTI then `br x30` will cause an exception as it jumps to a non landing pad instruction.
(We only generate landing pads for `return-twice` locations but a normal `catch` block won't get it)

Comment (john-brawn-arm) [reply libunwind/src/UnwindRegistersRestore.S:683]: Yes you're right, and actually the way I've adjusted __unw_phase2_resume also means we need to change the BTI instruction in __libunwind_Registers_arm64_jumpto to BTI JC. I'll fix this.

Comment (john-brawn-arm) [reply libunwind/src/UnwindRegistersRestore.S:683]: https://github.com/llvm/llvm-project/pull/102322 should fix this

**Pull request merged**